File: javascript/leetcode/349. Intersection of Two Arrays.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersection = function(nums1, nums2) {
  let st = new Set();
  let which = nums1.length > nums2.length ? nums1 : nums2;
  let other = nums1.length > nums2.length ? nums2 : nums1;

  which.forEach(el => {
    st.add(el);
  });

  let ans = [];
  other.forEach(el => {
    if (st.has(el)) {
      ans.push(el);
      st.delete(el);
    }
  });
  return ans;
};

File: go/leetcode/349. Intersection of Two Arrays.go
func intersection(nums1 []int, nums2 []int) []int {
    set := map[int]int{}
    which := []int{}
    other := which
    if len(nums1) > len(nums2) {
        which = nums1
        other = nums2
    } else {
        which = nums2
        other = nums1
    }
    
    for i := 0; i < len(which); i++ {
        set[which[i]] = 0
    }
    
    ans := []int{}
    
    for i := 0; i < len(other); i++ {
        _, ok := set[other[i]]
        if ok {
            ans = append(ans, other[i])
            delete(set, other[i])
        }
    }
    
    return ans
}